File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/mobs/spiders/SpiderExploding.java
package com.nyrds.pixeldungeon.mobs.spiders;

import com.nyrds.pixeldungeon.levels.objects.LevelObjectsFactory;
import com.nyrds.pixeldungeon.mobs.common.MultiKindMob;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.items.food.MysteryMeat;
import com.watabou.pixeldungeon.plants.Plant;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

public class SpiderExploding extends MultiKindMob {

    private static final String[] PlantClasses = {
            "Firebloom",
            "Icecap",
            "Sorrowmoss",
            "Dreamweed",
            "Sungrass",
            "Earthroot",
            "Fadeleaf",
            "Moongrace"
    };

    public SpiderExploding() {
        hp(ht(5));
        baseDefenseSkill = 1;
        baseAttackSkill  = 125;
        baseSpeed = 2f;
        dmgMin = 3;
        dmgMax = 6;
        dr = 0;

        expForKill = 3;
        maxLvl = 9;

        kind = Random.IntRange(0, 7);

        loot(new MysteryMeat(), 0.067f);
    }

    @Override
    public boolean attack(@NotNull Char enemy) {
        if (super.attack(enemy)) {

            Plant plant = (Plant) LevelObjectsFactory.objectByName(PlantClasses[getKind()]);
            plant.effect(enemy.getPos(), enemy);

            die(this);
            return true;
        }
        return false;
    }
}

File: scripts/mobs/SpiderExploding.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

-- java SpiderExploding (batch 12): living bomb - on a hit applies its random
-- plant to the victim and bursts (attackProc fires only on a hit, same as the
-- java attack() override that died on super.attack()==true). Java picked the
-- plant via a random MultiKindMob kind; kind is visually inert, so the roll
-- lives in lua data.
local PLANTS = {
    "Firebloom", "Icecap", "Sorrowmoss", "Dreamweed",
    "Sungrass", "Earthroot", "Fadeleaf", "Moongrace"
}

return mob.init{
    stats = function(self)
        local data = mob.restoreData(self)
        if data.plant then
            return
        end
        data.plant = PLANTS[math.random(#PLANTS)]
    end,
    attackProc = function(self, enemy, dmg)
        if enemy ~= nil then
            local plant = RPD.LevelObjectsFactory:objectByName(mob.restoreData(self).plant)
            if plant ~= nil then
                plant:effect(enemy:getPos(), enemy, self)
            end
            self:die(self)
        end
        return dmg
    end
}
